### PR TITLE
WIP: add support for private repos to open issues or wiki

### LINKED ISSUE #639: Public issue trackers and/or wikis for private repositories
I would really like it if I could set a repository's issue tracker and/or wiki to public, so that other people can create issues and look at/edit the wiki, even though the repository itself remains private. (I've been mainly using Bitbucket before switching to Gitea and I've always found that functionality very useful.)

Actually, what would be even better (though also probably a lot more work) is if there were seperate permissions for the issue tracker and wiki altogether, so that one could give certain people only read access to the wiki, others write access to the wiki, others access to the issue tracker, etc. I believe Gitlab has a system like that in place.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/40819688-public-issue-trackers-and-or-wikis-for-private-repositories?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by lunny**
This PR will fix #639. 
- [x] update repo settings
- [x] update permission check
- [ ] list repos who allow anonymous visit on explore

### CODE REVIEW

File: models/repo_permission.go [hunk 1]
```diff
@@ -23,25 +23,51 @@ func (p *Permission) IsAdmin() bool {
...
 		return AccessModeNone
 	}
 	return p.UnitsMode[unitType]
 }
 
+// CanAnonymousAccess returns true if anonymous access is enabled
+func (p *Permission) CanAnonymousAccess(unitType UnitType) bool {
+	for _, u := range p.Units {
+		if u.Type == unitType {
+			if u.AllowAnonymous {
```
Comment (lafriks) [inline models/repo_permission.go]: this `if` can be replaced with:
```go
return u.AllowAnonymous
```

File: options/locale/locale_en-US.ini [hunk 2]
```diff
@@ -1021,6 +1021,8 @@ settings.tracker_issue_style = External Issue Tracker Number Format
 settings.tracker_issue_style.numeric = Numeric
 settings.tracker_issue_style.alphanumeric = Alphanumeric
 settings.tracker_url_format_desc = Use the placeholders <code>{user}</code>, <code>{repo}</code> and <code>{index}</code> for the username, repository name and issue index.
+settings.anonymous_access_issues = Allow Anonymous Access Issues
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.anonymous_access_issues = Allow Anonymous Access to Issues
```

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -1021,6 +1021,8 @@ settings.tracker_issue_style = External Issue Tracker Number Format
...
+settings.anonymous_access_wiki = Allow Anonymous Access Wiki
```
Comment (lafriks) [inline options/locale/locale_en-US.ini]:
```suggestion
settings.anonymous_access_wiki = Allow Anonymous Access to Wiki
```

File: models/repo_permission.go (hunk 1 above)
Comment (lunny) [reply models/repo_permission.go]: done.

File: options/locale/locale_en-US.ini (hunk 2 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: options/locale/locale_en-US.ini (hunk 3 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: models/repo_permission.go [hunk 4]
```diff
@@ -23,25 +23,49 @@ func (p *Permission) IsAdmin() bool {
 
 // HasAccess returns true if the current user has at least read access to any unit of this repository
 func (p *Permission) HasAccess() bool {
+	for _, u := range p.Units {
+		if u.AllowAnonymous {
+			return true
```
Comment (adelowo) [inline models/repo_permission.go:28]: I think `return u.AllowAnonymous` should work here and the if block 
can be dropped

Comment (lafriks) [reply models/repo_permission.go:28]: I don't think that will be the same and will break logic

Comment (adelowo) [reply models/repo_permission.go:28]: Whoops, true.

Sorry, I have been writing too much Ruby of late where you 
could ___inline returns with an if block___

**Comment by EdaNano** [comment]
Hi, is there news about this merge? I would really like to see #639 resolved. Is this planned soon? 
Thank you :)

**Comment by lunny** [comment]
I may continue the work on this during v1.12.

**Comment by EdaNano** [comment]
Thank you for the quick response :)  When is v1.12. to come aprox? May means not sure, isn't it?

**Comment by FalconBFR** [comment]
Any ETA for this? Many thanks

**Comment by lunny** [comment]
@FalconBFR Sorry, it's not ready for v1.12.

**Comment by brunoais** [comment]
@lunny Is being ready for v1.13 very likely or is it more of a possible but not that likely? (Just trying to organize myself)

**Comment by coolguy1771** [comment]
Has there been any progress on this issue?

**Comment by wxiaoguang** [comment]
This feature would be very complex. For example: if a writer write a private repo URL in the the repo's public Issue, it would leak something in future, GitHub can show the code referred in issue.

In my environment, I just split repositories into two, one for private code, one for public issue/wiki.

**Comment by lunny** [comment]
Closed as it's outdated.

**Pull request closed without merging**